### DISCUSSION TITLE
Concentration and Temperature Dependent NeumannBC

### DISCUSSION THREAD

**Discussion by styyokuda** [idaholab/moose#17869] [Q&A General] [answered]
Dear MOOSE Team,
For my diffusion and heat model, I am writing a NeumannBC that is concentration and temperature dependent (please see the attached files for a simplified version).
Just for the simplification, I just write here as:
return -_test[_i][_qp]*_u[_qp]*_temperature[_qp]/273.15*_normals[_qp];

It compiles without "_normals[_qp]."
Also in the input file, "variable = u" needs to be set as:
[BCs]
  [./right_u]
    type = AnotherNeumannBC
    variable = u
    diffusion_variable = u
    heat_variable = temperature
    boundary = 'right'
  [../]
[ ]

It would be appreciated if anyone could explain why "_normals[_qp]" cannot be used 
and "variable = u" needs to be set (I think "diffusion_variable = u" is enough).
Please help me.
Thomas

If "_normals[_qp]" is included, the error message is:
~/AnotherNeumannBC.C:39:10: error: no viable conversion from returned value of type 'typename
      boostcopy::enable_if_c<ScalarTraits<double>::value, TypeVector<typename CompareTypes<double, double>::supertype> >::type'
      (aka 'libMesh::TypeVector<double>') to function return type 'libMesh::Real' (aka 'double')
  return -_test[_i][_qp]*_u[_qp]*_temperature[_qp]/273.15*_normals[_qp];
        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
 
[AnotherNeumannBC.i.docx](https://github.com/idaholab/moose/files/6498427/AnotherNeumannBC.i.docx)
[AnotherNeumannBC.h.docx](https://github.com/idaholab/moose/files/6498434/AnotherNeumannBC.h.docx)
[AnotherNeumannBC.C.docx](https://github.com/idaholab/moose/files/6498435/AnotherNeumannBC.C.docx)

**Comment by dschwen** [comment, marked as answer]
The BC residual needs to be a scalar, you're retuning a vector. Look at the
weak form derivation. The normal does not show up in it. You're probably
thinking intuitively there should be a gradient projected along the normal
(only you don't have a gradient in there either).

**Reply by styyokuda** [reply]
Thank you very much for a great explanation.

Two more questions please.
(1) In the input, "variable = u" is required because "variable" is a required parameter for the "NeumannBC" object requires (so, I don't need to setup "diffusion_variable = u" here).
(2) I modified the "NeumannBC" object here but, more complicated calculations, I can use the "Material" object, too.
Please let me know if my understanding of the above two is correct.
Thomas

**Reply by GiudGiud** [reply]
1) indeed. No need to specify the variable twice if it's clear what's the variable
2) You can use Material objects in a modified NeumannBC, just inherit a MaterialPropertyInterface in your BC object
